### PR TITLE
Add the cdn edit option.

### PR DESCRIPTION
**Pull request by FernandoOjeda**
Add the cdn edit option.
#1471

### CODE REVIEW

File: SoftLayer/CLI/cdn/edit.py
Comment (caberos) [inline SoftLayer/CLI/cdn/edit.py:73]: `CacheKeyQueryRule`  change the repository format

Comment (caberos) [reply SoftLayer/CLI/cdn/edit.py:73]: please review the table format and changes to repository format

File: SoftLayer/managers/cdn.py
Comment (allmightyspiff) [inline SoftLayer/managers/cdn.py:241]: I'd recommend making both these strings lowercase when doing a check
```suggestion
            if cdn.get('domain', '').lower() == hostname.lower():
```

File: SoftLayer/managers/cdn.py
Comment (allmightyspiff) [inline SoftLayer/managers/cdn.py:257]: make this 1 line.

**Review by allmightyspiff — CHANGES_REQUESTED** [review]
looks pretty good, just a few minor changes needed. Thanks.

File: SoftLayer/CLI/cdn/edit.py
Comment (FernandoOjeda) [reply SoftLayer/CLI/cdn/edit.py:51]: Done.

File: SoftLayer/managers/cdn.py
Comment (FernandoOjeda) [reply SoftLayer/managers/cdn.py]: Done.

File: SoftLayer/CLI/cdn/edit.py
Comment (FernandoOjeda) [reply SoftLayer/CLI/cdn/edit.py]: Done.

File: SoftLayer/managers/cdn.py
Comment (FernandoOjeda) [reply SoftLayer/managers/cdn.py]: Done.

File: SoftLayer/managers/cdn.py
Comment (FernandoOjeda) [reply SoftLayer/managers/cdn.py:257]: Done.

**Review by allmightyspiff — APPROVED** [review]
Looks good to me, would like someone to post some tests results for this one though before I merge it.

Thanks.

**Comment by FernandoOjeda** [comment]
## COMMANDS TESTED:

+ ✔️  slcli cdn edit --origin 10.32.12.125 --http-port 83 1111111111
    ```
    Calling: SoftLayer_Network_CdnMarketplace_Configuration_Mapping::listDomainMappingByUniqueId(id=None, mask='', filter='None', args=('1111111111111',), limit=None, offset=None))
    Calling: SoftLayer_Network_CdnMarketplace_Configuration_Mapping::updateDomainMapping(id=None, mask='', filter='None', args=({'uniqueId': '111111111111', 'originType': 'HOST_SERVER', 'protocol': 'H
TTP', 'path': '/*', 'vendorName': 'akamai', 'cname': 'test': 'www.test.com', 'httpPort': 83, 'origin': '10.32.12.125'},), limit=None, offset=None
))
    ```

+ ✔️  slcli -v cdn edit --origin 10.32.12.125 --http-port 82 www.test.com
    ```
    Calling: SoftLayer_Network_CdnMarketplace_Configuration_Mapping::listDomainMappings(id=None, mask='', filter='None', args=(), limit=None, offset=None))
    Calling: SoftLayer_Network_CdnMarketplace_Configuration_Mapping::listDomainMappingByUniqueId(id=None, mask='', filter='None', args=('11111111111',), limit=None, offset=None))
    Calling: SoftLayer_Network_CdnMarketplace_Configuration_Mapping::updateDomainMapping(id=None, mask='', filter='None', args=({'uniqueId': '111111111111', 'originType': 'HOST_SERVER', 'protocol': 'H
TTP', 'path': '/*', 'vendorName': 'akamai', 'cname': 'test', 'domain': 'www.test.com', 'httpPort': 82, 'origin': '10.32.12.125'},), limit=None, offset=None
))
    ```

+ ✔️ slcli -v cdn edit --origin 10.32.12.125 --header www.test2.com  11111111111
    ```
    Calling: SoftLayer_Network_CdnMarketplace_Configuration_Mapping::listDomainMappingByUniqueId(id=None, mask='', filter='None', args=('1111111111',), limit=None, offset=None))
    Calling: SoftLayer_Network_CdnMarketplace_Configuration_Mapping::updateDomainMapping(id=None, mask='', filter='None', args=({'uniqueId': '1111111111', 'originType': 'HOST_SERVER', 'protocol': 'H
TTP', 'path': '/*', 'vendorName': 'akamai', 'cname': 'test', 'domain': 'www.test.com', 'httpPort': 82, 'header': 'www.test2.com', 'origin': '10.32.12.125'}
,), limit=None, offset=None))
    ```

+ ✔️ slcli -v cdn edit --origin 10.32.12.125 --header www.test2.com  www.test.com
    ```
    Calling: SoftLayer_Network_CdnMarketplace_Configuration_Mapping::listDomainMappingByUniqueId(id=None, mask='', filter='None', args=('1111111111',), limit=None, offset=None))
    Calling: SoftLayer_Network_CdnMarketplace_Configuration_Mapping::updateDomainMapping(id=None, mask='', filter='None', args=({'uniqueId': '1111111111', 'originType': 'HOST_SERVER', 'protocol': 'H
TTP', 'path': '/*', 'vendorName': 'akamai', 'cname': 'test', 'domain': 'www.test.com', 'httpPort': 82, 'header': 'www.test2.com', 'origin': '10.32.12.125'}
,), limit=None, offset=None))
    ```

+ ✔️ slcli -v cdn edit --origin 10.32.12.125 --respect-headers 1  11111111111
    ```
    Calling: SoftLayer_Network_CdnMarketplace_Configuration_Mapping::listDomainMappingByUniqueId(id=None, mask='', filter='None', args=('11111111111',), limit=None, offset=None))
    Calling: SoftLayer_Network_CdnMarketplace_Configuration_Mapping::updateDomainMapping(id=None, mask='', filter='None', args=({'uniqueId': '1111111111111', 'originType': 'HOST_SERVER', 'protocol': 'H
TTP', 'path': '/*', 'vendorName': 'akamai', 'cname': 'testd', 'domain': 'www.test.com', 'httpPort': 82, 'origin': '10.32.12.125', 'respectHeaders': '1'},),
limit=None, offset=None))
    ```

+ ✔️ slcli -v cdn edit --origin 10.32.12.125 --respect-headers 1  www.test.com
    ```
    Calling: SoftLayer_Network_CdnMarketplace_Configuration_Mapping::listDomainMappingByUniqueId(id=None, mask='', filter='None', args=('11111111111',), limit=None, offset=None))
    Calling: SoftLayer_Network_CdnMarketplace_Configuration_Mapping::updateDomainMapping(id=None, mask='', filter='None', args=({'uniqueId': '1111111111111', 'originType': 'HOST_SERVER', 'protocol': 'H
TTP', 'path': '/*', 'vendorName': 'akamai', 'cname': 'testd', 'domain': 'www.test.com', 'httpPort': 82, 'origin': '10.32.12.125', 'respectHeaders': '1'},),
limit=None, offset=None))
    ```

+ ✔️ slcli -v cdn edit --origin 10.32.12.125 --cache include-all  www.test.com
    ```
    Calling: SoftLayer_Network_CdnMarketplace_Configuration_Mapping::listDomainMappings(id=None, mask='', filter='None', args=(), limit=None, offset=None))
    Calling: SoftLayer_Network_CdnMarketplace_Configuration_Mapping::listDomainMappingByUniqueId(id=None, mask='', filter='None', args=('111111111',), limit=None, offset=None))
    Calling: SoftLayer_Network_CdnMarketplace_Configuration_Mapping::updateDomainMapping(id=None, mask='', filter='None', args=({'uniqueId': '11111111111', 'originType': 'HOST_SERVER', 'protocol': 'H
TTP', 'path': '/*', 'vendorName': 'akamai', 'cname': 'test', 'domain': 'www.test.com', 'httpPort': 82, 'origin': '10.32.12.125', 'cacheKeyQueryRule': 'incl
ude-all'},), limit=None, offset=None))
    ```

+ ✔️ slcli -v cdn edit --origin 10.32.12.125 --cache include-all  www.test.comslcli -v cdn edit --origin 10.32.12.125 --cache include-specified --cache test 11111111111
    ```
    Calling: SoftLayer_Network_CdnMarketplace_Configuration_Mapping::listDomainMappingByUniqueId(id=None, mask='', filter='None', args=('11111111111',), limit=None, offset=None))
    Calling: SoftLayer_Network_CdnMarketplace_Configuration_Mapping::updateDomainMapping(id=None, mask='', filter='None', args=({'uniqueId': '1111111111111', 'originType': 'HOST_SERVER', 'protocol': 'H
TTP', 'path': '/*', 'vendorName': 'akamai', 'cname': 'test', 'domain': 'www.test.com', 'httpPort': 82, 'origin': '10.32.12.125', 'cacheKeyQueryRule': 'incl
ude: test'},), limit=None, offset=None))
    ```

+ ✔️ slcli -v cdn edit --origin 10.32.12.125 --cache include-all  www.test.comslcli -v cdn edit --origin 10.32.12.125 --cache include-specified --cache test 11111111111
    ```
    Calling: SoftLayer_Network_CdnMarketplace_Configuration_Mapping::listDomainMappingByUniqueId(id=None, mask='', filter='None', args=('11111111111',), limit=None, offset=None))
    Calling: SoftLayer_Network_CdnMarketplace_Configuration_Mapping::updateDomainMapping(id=None, mask='', filter='None', args=({'uniqueId': '1111111111111', 'originType': 'HOST_SERVER', 'protocol': 'H
TTP', 'path': '/*', 'vendorName': 'akamai', 'cname': 'test', 'domain': 'www.test.com', 'httpPort': 82, 'origin': '10.32.12.125', 'cacheKeyQueryRule': 'incl
ude: test'},), limit=None, offset=None))
    ```

File: SoftLayer/CLI/cdn/edit.py
Comment (ATGE) [inline SoftLayer/CLI/cdn/edit.py:79]: Please add a field for `Origin server address`

**Review by ATGE — CHANGES_REQUESTED** [review]
### Testing
- ✔️ slcli cdn edit www.techsuppor4.com --origin=10.10.10.5 --header=www.techsuppot-test.com  --http-port=80
-  ❌slcli cdn edit www.techsuppor4.com --origin=10.10.10.10
 
image

-  ❌slcli cdn edit www.techsuppor4.com  --header=www.techsuppot-test2.com 
 
image

-  ❌slcli cdn edit www.techsuppor4.com --http-port=80
 
image

File: SoftLayer/CLI/cdn/edit.py
Comment (ATGE) [reply SoftLayer/CLI/cdn/edit.py:23]: image
I can able to edit just one field, so there is no mandatory edit `Origin server address`.

Comment (FernandoOjeda) [reply SoftLayer/CLI/cdn/edit.py:23]: Did you try to remove the data from the `Origin server address` field, you will see that the save button is not activated lol 
image
.

image

**Comment by FernandoOjeda** [comment]
The `Hostname` and the `Host header` can have the same data lol.
You can change the `Host header` data if you want. 
image

image

**Comment by FernandoOjeda** [comment]
The `Origin server address` is required by UI and API.

If you remove the data from the `Origin server address` field, you will see that the save button is not activated lol. 
image

image

This is the request that the control portal uses, needs origin data.
image

By rest call:
image

**Comment by ATGE** [comment]
> The `Hostname` and the `Host header` can have the same data lol.
> You can change the `Host header` data if you want.
> image
> 
> image

I agree that can be the same, but in `slcli cdn edit www.techsuppor4.com --origin=10.10.10.10`, just want to change the origin, not the header too.

File: SoftLayer/CLI/cdn/edit.py
Comment (ATGE) [reply SoftLayer/CLI/cdn/edit.py:23]: should edit just header I think, and take the current origin as UI make, if I do not want to change it an explicit way.
I agree with the origin parameter is mandatory in API call, but at the command level, if the customer does not desire to edit the current one, the command could use the current data as default, and just edit the explicit parameter, header in this case, as UI do it.

**Comment by FernandoOjeda** [comment]
If you see your screenshot, it shows other thing, please be more clear.

image

This command is working for me:
`slcli cdn edit www.4techsupporttest.com --origin 10.10.10.10`

image

File: SoftLayer/CLI/cdn/edit.py
Comment (ATGE) [reply SoftLayer/CLI/cdn/edit.py:23]: about `slcli cdn edit www.techsuppor4.com --origin=10.10.10.10` updates the origin server which it's right, but also is using the identifier as header parameter. so I think is a side effect.

Comment (FernandoOjeda) [reply SoftLayer/CLI/cdn/edit.py:23]: I recommend you to check well the UI part of cdn settings and the documentation, you can edit the header field with the data you want, but you can not edit the hostname lol.

image

###Host header
`The edge server uses the host header in the HTTP header to communicate with the Origin host. It defaults to Hostname.`

**Comment by allmightyspiff** [comment]
> I agree that can be the same, but in `slcli cdn edit www.techsuppor4.com --origin=10.10.10.10`, just want to change the origin, not the header too.

I don't think this command is actually changing the origin.


```bash
./slcli -v  cdn edit www.techsuppor4.com --origin 10.10.10.10
Calling: SoftLayer_Network_CdnMarketplace_Configuration_Mapping::listDomainMappings(id=None, mask='', filter='None', args=(), limit=None, offset=None))
Calling: SoftLayer_Network_CdnMarketplace_Configuration_Mapping::listDomainMappingByUniqueId(id=None, mask='', filter='None', args=('507616172924769',), limit=None, offset=None))
Calling: SoftLayer_Network_CdnMarketplace_Configuration_Mapping::updateDomainMapping(id=None, mask='', filter='None', args=({'uniqueId': '507616172924769', 'originType': 'HOST_SERVER', 'protocol': 'HTTP', 'path': '/example2test/*', 'vendorName': 'akamai', 'cname': 'cdnakaog136c8gfq12221.cdn.appdomain.cloud', 'domain': 'www.techsuppor4.com', 'httpPort': 81, 'origin': '10.10.10.10'},), limit=None, offset=None))

SoftLayerAPIError(500): Status [500]: Update domain mapping failed with unexpected error. Message: Status[500]: Update Domain Mapping failed at vendor layer
```

Compared to 

```
slcli -v  cdn edit www.techsuppor4.com --origin 10.10.10.10 -H www.techsupport4.com
Calling: SoftLayer_Network_CdnMarketplace_Configuration_Mapping::listDomainMappings(id=None, mask='', filter='None', args=(), limit=None, offset=None))
Calling: SoftLayer_Network_CdnMarketplace_Configuration_Mapping::listDomainMappingByUniqueId(id=None, mask='', filter='None', args=('507616172924769',), limit=None, offset=None))
Calling: SoftLayer_Network_CdnMarketplace_Configuration_Mapping::updateDomainMapping(id=None, mask='', filter='None', args=({'uniqueId': '507616172924769', 'originType': 'HOST_SERVER', 'protocol': 'HTTP', 'path': '/example2test/*', 'vendorName': 'akamai', 'cname': 'cdnakaog136c8gfq12221.cdn.appdomain.cloud', 'domain': 'www.techsuppor4.com', 'httpPort': 81, 'header': 'www.techsupport4.com', 'origin': '10.10.10.10'},), limit=None, offset=None))
SoftLayerAPIError(500): Status [500]: Update domain mapping failed with unexpected error. Message: Status[500]: Update Domain Mapping failed at vendor layer

```

I'm not sure why were getting a 500 here, but I don't think its a SL API problem, maybe the CDN account is just in an odd status or something?

**Comment by FernandoOjeda** [comment]
@allmightyspiff The rest of the slcli cdn edit commands are working for you? because all of them are working for me, it is rare that this error appears for you, since we are not changing the domain data.

This is working for rest:
image


`slcli cdn edit www.1techsupporttest.com --origin 10.34.12.125 -H www.1techsupporttest.com`

image

`slcli cdn edit www.1techsupporttest.com --origin 10.10.10.10`

image

It working for xmlrcp too:
image

image

**Comment by allmightyspiff** [comment]
The command work properly now:

image

Given the error, I was assuming it was a problem on the CDN side of things, not with your code.

@ATGE  when you get a moment can you rereview this and approve/request changes if needed. Thanks.

File: SoftLayer/CLI/cdn/edit.py
Comment (allmightyspiff) [reply SoftLayer/CLI/cdn/edit.py:23]: Looks like Origin is required in the UI, so having it required here makes sense.

File: SoftLayer/CLI/cdn/edit.py
Comment (allmightyspiff) [reply SoftLayer/CLI/cdn/edit.py:79]: @FernandoOjeda Could you add this field. Once that is added I think this issue should be ready to go.

File: SoftLayer/managers/cdn.py
Comment (ATGE) [inline SoftLayer/managers/cdn.py:203]:
```suggestion
            'origin': cdn_instance_detail.get('originHost')
            'header': cdn_instance_detail.get('header')
```
Please add `header` as a default value too.

**Review by ATGE — CHANGES_REQUESTED** [review]
looks good, just a comment to get the header as default value too, thanks.

File: SoftLayer/managers/cdn.py
Comment (FernandoOjeda) [reply SoftLayer/managers/cdn.py:203]: What is the purpose of adding the `header` data here for updating the cdn object?

Comment (ATGE) [reply SoftLayer/managers/cdn.py:203]: if the` header` is here, this one just will update if is passed as explicit param. so the header will behavior, in the same way, the `origin`, and the `httpPort` are working.

Comment (FernandoOjeda) [reply SoftLayer/managers/cdn.py]: Done, thanks.

**Review by ATGE — APPROVED** [review]
Changes look good, thanks!

### Testing
- ✔️ slcli cdn edit www.techsuppor4.com --http-port=81
- ✔️ slcli cdn edit www.techsuppor4.com --header=www.techsuppot-test2.com
- ✔️ slcli cdn edit 507616172924769 --header=www.techsuppot-test3.com
- ✔️ slcli cdn edit www.techsuppor4.com --origin=10.10.10.10
- ✔️ slcli -v cdn edit www.techsuppor4.com --origin=10.10.10.5 --header=www.techsuppot-test.com --http-port=80

**Pull request merged**
